### PR TITLE
Cookies #23dq8hb

### PR DESCRIPTION
**Pull request by sdimitris**
- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other (please describe)

**Does this PR introduce a breaking change?** 

- [ ] Yes
- [x] No

If adding a **new feature**, the PR's description includes:

Many applications use cookies for their authentication tokens, and conduit should have an option to set cookies instead of providing creds in response.

- Cookie functionality is configurable through `setCookies.enabled` config option.
- When cookies are returned from authentication module, router wraps them in the response object.
- When a request arrives, cookies are accessible through `call.request.context.cookies` object.

### CODE REVIEW

File: libraries/grpc-sdk/src/helpers/RoutingUtilities.ts [hunk 1]
```diff
@@ -22,6 +22,8 @@ message RouterRequest {
 message RouterResponse {
   string result = 1;
   string redirect = 2;
+  optional string setCookies = 3;
```
Comment (kon14) [inline libraries/grpc-sdk/src/helpers/RoutingUtilities.ts:25]: We [don't](https://github.com/ConduitPlatform/Conduit/issues/86) currently follow this yet, but gRPC's style guide suggests that we [use underscore_separated_names for message field names](https://developers.google.com/protocol-buffers/docs/style#message_and_field_names).

File: modules/authentication/src/handlers/service.ts [hunk 2]
```diff
@@ -65,9 +66,11 @@ export class ServiceHandler {
         userId: serviceUser._id,
         clientId: context.clientId,
         config,
-      }
+      },
     );
-
+    if (config.set_cookies.enabled) {
+      return AuthUtils.returnCookies('setCookies','Successfully authenticated',(accessToken as any).token,(refreshToken as any).token)
```
Comment (kon14) [inline modules/authentication/src/handlers/service.ts:72]: Cast these to `string` instead.

File: modules/authentication/src/handlers/AuthenticationProviders/OAuth2.ts [hunk 3]
```diff
@@ -112,6 +112,10 @@ export abstract class OAuth2<T extends Payload, S extends OAuth2Settings> {
     let user = await this.createOrUpdateUser(payload);
     const config = ConfigController.getInstance().config;
     let tokens = await this.createTokens(user._id, call.request.params['clientId'], config);
+
+    if (config.set_cookies.enabled) {
+      return AuthUtils.returnCookies('setCookies','Successfully authenticated',(tokens.accessToken as any),(tokens.refreshToken as any))
```
Comment (kon14) [inline modules/authentication/src/handlers/AuthenticationProviders/OAuth2.ts:117]: Cast these to `string` instead.

**Review by kon14 — CHANGES_REQUESTED** [review]

File: libraries/grpc-sdk/src/helpers/RoutingUtilities.ts [hunk 4]
```diff
@@ -31,6 +33,18 @@ message SocketRequest {
   string context = 4;
 }
 
+message Cookie {
```
Comment (kkopanidis) [inline libraries/grpc-sdk/src/helpers/RoutingUtilities.ts:36]: Missing SameSite, Domain and Path options

File: libraries/grpc-sdk/src/helpers/wrapRouterFunctions.ts [hunk 5]
```diff
@@ -68,14 +68,23 @@ export function wrapRouterGrpcFunction(
           if (typeof r === 'string') {
             callback(null, { result: r });
           } else {
+            const result = r.result ? JSON.stringify(r.result) : undefined;
```
Comment (kkopanidis) [inline libraries/grpc-sdk/src/helpers/wrapRouterFunctions.ts:71]: Do not modify the order in which parsing and checks happen. Add an "if" statement on the top, when you create a "response" object that contains the cookie options. After that the rest of the checks should add their data in that object. For example, line 69 would do: callback(null, {...responseObject,result:r}

File: modules/authentication/src/handlers/AuthenticationProviders/OAuth2.ts [hunk 6]
```diff
@@ -112,10 +113,30 @@ export abstract class OAuth2<T extends Payload, S extends OAuth2Settings> {
...
+        options: cookieOptions,
+      }];
+      if (!isNil(tokens.refreshToken)) {
+        cookies.push({
+          name: 'refreshToken',
+          value: tokens.refreshToken,
+          options: cookieOptions,
+        });
+      }
+      return {
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/AuthenticationProviders/OAuth2.ts:130]: message is not needed, return the userId as we regularly do and setCookies

File: modules/authentication/src/handlers/facebook/facebook.ts [hunk 7]
```diff
@@ -93,8 +93,11 @@ export class FacebookHandlers extends OAuth2<Payload, FacebookSettings> {
       },
       new ConduitRouteReturnDefinition('FacebookResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/facebook/facebook.ts:98]: set/removeCookies is not a response as I've said yesterday. Remove it and also remove the message.

File: modules/authentication/src/handlers/figma/figma.ts [hunk 8]
```diff
@@ -81,8 +81,11 @@ export class FigmaHandlers extends OAuth2<FigmaUser, FigmaSettings> {
       },
       new ConduitRouteReturnDefinition('FigmaResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/figma/figma.ts:86]: same as my previous comment

File: modules/authentication/src/handlers/github/github.ts [hunk 9]
```diff
@@ -69,8 +69,11 @@ export class GithubHandlers extends OAuth2<GithubUser, GithubSettings> {
       },
       new ConduitRouteReturnDefinition('GithubResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/github/github.ts:74]: same as my previous comment

File: modules/authentication/src/handlers/google/google.ts [hunk 10]
```diff
@@ -103,8 +103,11 @@ export class GoogleHandlers extends OAuth2<GoogleUser, GoogleSettings> {
       },
       new ConduitRouteReturnDefinition('GoogleResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/google/google.ts:108]: same as my previous comment

File: modules/authentication/src/handlers/local.ts [hunk 11]
```diff
@@ -200,20 +201,43 @@ export class LocalHandlers {
...
+        result: { message: 'Successfully authenticated' },
+        setCookies: cookies,
+      };
+    }
 
     return {
       userId: user._id.toString(),
       accessToken: accessToken.token,
-      refreshToken: refreshToken.token,
+      refreshToken: !isNil(refreshToken!) ? refreshToken!.token : undefined,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/local.ts:240]: you're checking if the refreshToken is null, and at the same time you're using "refreshToken!" these are contradictory.

File: modules/authentication/src/handlers/microsoft/microsoft.ts [hunk 12]
```diff
@@ -73,8 +73,11 @@ export class MicrosoftHandlers extends OAuth2<MicrosoftUser, MicrosoftSettings>
       },
       new ConduitRouteReturnDefinition('MicrosoftResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/microsoft/microsoft.ts:78]: you know what

File: modules/authentication/src/handlers/phone.ts [hunk 13]
```diff
@@ -131,20 +132,40 @@ export class PhoneHandlers {
...
+        return {
+          result: { message: 'Successfully authenticated' },
+          setCookies: cookies,
+        };
+      }
       return {
         userId: user._id.toString(),
         accessToken: accessToken.token,
-        refreshToken: refreshToken.token,
+        refreshToken: !isNil(refreshToken!) ? (refreshToken as RefreshToken).token : undefined,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/phone.ts:168]: same

File: modules/authentication/src/handlers/service.ts [hunk 14]
```diff
@@ -65,13 +67,31 @@ export class ServiceHandler {
...
+      return {
+        result: { message: 'Successfully authenticated' },
+        setCookies: cookies,
+      };
+    }
     return {
       serviceId: serviceUser._id.toString(),
       accessToken: (accessToken as any).token,
-      refreshToken: (refreshToken as any).token,
+      refreshToken: !isNil(refreshToken!) ? (refreshToken as RefreshToken).token : undefined,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/service.ts:94]: same

File: modules/authentication/src/handlers/slack/slack.ts [hunk 15]
```diff
@@ -82,8 +82,11 @@ export class SlackHandlers extends OAuth2<SlackUser, SlackSettings> {
       },
       new ConduitRouteReturnDefinition('SlackResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
+        removeCookies: { type: [TYPE.String], required: false },
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/slack/slack.ts:88]: same as my previous comment

File: modules/authentication/src/handlers/twitch/twitch.ts [hunk 16]
```diff
@@ -66,8 +66,11 @@ export class TwitchHandlers extends OAuth2<TwitchUser, TwitchSettings> {
       },
       new ConduitRouteReturnDefinition('TwitchResponse', {
         userId: ConduitString.Required,
-        accessToken: ConduitString.Required,
-        refreshToken: ConduitString.Required,
+        accessToken: ConduitString.Optional,
+        refreshToken: ConduitString.Optional,
+        message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/handlers/twitch/twitch.ts:71]: same as my previous comment

File: modules/authentication/src/interfaces/Cookie.ts [hunk 17]
```diff
@@ -0,0 +1,10 @@
+export interface Cookie {
+  name: string;
+  value: string;
+  options?: {
+    httpOnly?: boolean;
```
Comment (kkopanidis) [inline modules/authentication/src/interfaces/Cookie.ts:5]: add the missing options

File: modules/authentication/src/routes/routes.ts [hunk 18]
```diff
@@ -108,14 +109,19 @@ export class AuthenticationRoutes {
           accessToken: ConduitString.Optional,
           refreshToken: ConduitString.Optional,
           message: ConduitString.Optional,
+          removeCookies: { type: [TYPE.String], required: false },
```
Comment (kkopanidis) [inline modules/authentication/src/routes/routes.ts:112]: nope

File: modules/authentication/src/routes/routes.ts [hunk 19]
```diff
@@ -332,6 +338,9 @@ export class AuthenticationRoutes {
           serviceId: ConduitString.Required,
           accessToken: ConduitString.Required,
           refreshToken: ConduitString.Required,
+          message: ConduitString.Optional,
```
Comment (kkopanidis) [inline modules/authentication/src/routes/routes.ts:341]: nope

File: packages/router/src/controllers/GraphQl/GraphQL.ts [hunk 20]
```diff
@@ -386,6 +388,15 @@ export class GraphQLController extends ConduitRouter {
         })
         .then((r) => {
           let result = r.result ? r.result : r;
+          if (r.setCookies) {
+            const cookies = JSON.parse(r.setCookies);
```
Comment (kkopanidis) [inline packages/router/src/controllers/GraphQl/GraphQL.ts:392]: parsing will fail. setCookies is an object in grpc not a string

**Review by kkopanidis — CHANGES_REQUESTED** [review]

File: packages/router/src/controllers/GraphQl/utils/cookie.plugin.ts [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+        const { setCookie, removeCookie } = requestContext.context;
+        const { res } = requestContext.context;
+
+        setCookie.forEach((cookie: any) => {
+          res.cookie(cookie.name, cookie.value, cookie.options);
+        });
+
+        if (removeCookie) {
+          removeCookie.forEach((cookie: any) => {
+            res.clearCookie(cookie.name, { domain: cookie.options.domain, path: cookie.options.path });
```
Comment (kkopanidis) [inline packages/router/src/controllers/GraphQl/utils/cookie.plugin.ts:17]: it needs all the options, not only domain and path

**Review by kkopanidis — CHANGES_REQUESTED** [review]

**Review by kkopanidis — APPROVED** [review]

**Review by kon14 — APPROVED** [review]

**Pull request merged**